chore: support for passing file contents for sourcemaps; puya/teal edge cases

aorumbayev opened the pull request:
- hidiing locals when debuggin TEAL 
- ignoring synthetic sources (see TODO)
- hiding 'op' in program state when its unknown

aorumbayev committed:
chore: hiding locals on non puya sourcemaps; adding puya and teal example

aorumbayev committed:
chore: no tmpl vars offset edge case fix

aorumbayev committed:
chore: rel path for example on puya + teal

aorumbayev committed:
chore: workaround on skipping syntentic refs that does not require async file exists

aorumbayev committed:
feat: support direct file or encoded content for sourcemaps

aorumbayev commented:
@daniel-makerx will add more refined puya example workspaces tomorrow in separate pr

daniel-makerx committed:
refactor: make programSourcesDescription an object instead of JSON

daniel-makerx commented on src/common/debugSession.ts:
```diff
@@ -42,7 +43,11 @@ export interface ILaunchRequestArguments
-  programSourcesDescriptionFile: string;
+  programSourcesDescriptionFile?: string;
+  /** JSON encoded content of the program sources description file. */
+  programSourcesDescription?: string;
```
This should be an object of type `ProgramSourceEntryFile` rather than the JSON encoded equivalent

daniel-makerx commented on src/common/programReplay.ts:
```diff
@@ -15,6 +16,7 @@ const HIDE_VERSION = true;
 const DEFINED_ONLY = true;
 const INCLUDE_STACK_SCOPE = false;
 const HIDE_TEMP = true;
+const PUYA_SOURCEMAPS_REGEX_BLOCKLIST = [/[/\\]algopy[/\\][^/\\]+\.py$/];
```
This would need to accomodate all potential frontend synthetic methods, instead I think this can be resolved in puya itself., by not outputting a location for synthetic methods

daniel-makerx commented on src/common/programReplay.ts:
```diff
@@ -175,6 +183,17 @@ export class ProgramReplay {
+    // which will require a lot of changes.
+    // As a simpler workaround adding a 'blocked' list of regexes for paths considered synthetic
+    // in puya source maps.
+    if (PUYA_SOURCEMAPS_REGEX_BLOCKLIST.some((regex) => source.match(regex))) {
```
See above comment on how we can resolve this for synthetic methods, however another likely source for missing locations will be third-party libraries. As when debugging the contract source might be available but not any libraries it was using.

Will investigate if there is another way to address this particular issue

aorumbayev committed:
chore: adding more edge cases

aorumbayev commented:
@daniel-makerx added the following cases to sample workspaces:
- puya -> claim_poa method simulation of a proof of attendance contract from unit testing repo
- puya_and_teal -> an intermix starting with an app call txn signed by delegated lsig (puya based) -> hello world app (teal based) sending itxn to another app -> hello world 2 app (puya based)

aorumbayev committed:
chore(src/common): Add program source entry file and type

Added 'ProgramSourceEntryFile' and 'ProgramSourceEntry' to the common module.
These types are used for parsing program source files.

aorumbayev committed:
chore: add support for 'null' sourcemap paths when user prefers ignores

aorumbayev committed:
chore: add mocharc file for global timeouts

Also fixing a few tests

aorumbayev commented:
@daniel-makerx added exports for ProgramSourceEntryFile and ProgramSourceEntry so the algokit extension can reuse them as a consumer. 

Also fixed a few unit tests but there are around 6-7 that are still failing, likely caused by changes in forward method, mocha console output is not helpful at all so need some time on monday to find root causes - alternatively we can fix those after deciding on how to deal with synthetics. Lets sync up more on this on monday.

daniel-makerx committed:
fix: move program forward call after inner transaction processing

aorumbayev committed:
chore: further refine puya example with third parties

aorumbayev committed:
feat: integrating Locals with existing expand avm methods

aorumbayev committed:
chore: adding puya unit test; moving process unit into program replay
